### ISSUE TITLE
Offline mode/Caching

### ISSUE THREAD

**Issue by fajabird** [marius-wieschollek/passwords-webextension#162]
**Current Status**
Currently the plugin does not have a local copy/cache of the passwords. This limits nextcloud password usage, e.g. application specific passwords that also need to be available when offline (e.g. laptop on the go). As there is no desktop app for nextcloud passwords as well the browser plugin is the only way for easy access to the passwords. I currently cannot recommend to use nextcloud passwords for other than some typical online websites (amazon) - but not for anything that is more critical (in terms of availability). While on server side nextcloud passwords shine (with e.g. sharing to other users) it really falls behind on any endpoints, incl. mobiles where one app has caching but no auto-fill and the other has auto-fill but no caching. 

**Feature Description**
Plugin stores and keeps up-to-date a local copy of the database. This not only speeds up the access to passwords but it also provides high availability e.g. when nextcloud is unavailable (a failed update "again"), network issues (laptop on the go) etc. 

**Additional context**

**Comment by marius-wieschollek** [comment]
This feature is a potential security risk. An unencrypted copy of the password database could be read by everyone with access to the computer. This also means that in case a device is stolen, revoking the credentials for the app isn't enough and the data is still left on the device with no way to delete it.

So if this is added, there needs to be a per account encryption for the saved data.

**Comment by fajabird** [comment]
Yes thats right - if a user opts in for caching he must be prompted for some additional credentials like a master password that will encrypt the offline cache. In principle it becomes a Keepass-like application - so therefore its not a disadvantage either, as long we keep it optional.

**Comment by cptMikky** [comment]
I would argue that the security implications are not as serious as stated. There is in fact an unencrypted cache present on the computer already, albeit _only_ in RAM which, arguably, _could be read by everyone with access to the computer_.

I have multiple issues with the claim that _"the password database could be read by everyone with access to the computer"_. First off, who is _everyone_? Is it another user with a separate user account? Is it someone with off-line physical access? Is it someone with remote access? Is it someone with user's access credentials? Is it something else completely?

Second: what is _access_?  Is it user access? If so, which user? Is it root access? Local? Remote? Off-line? Is it access to the hard drive only?

Each of these threat models have different impacts and different solutions. Not all of them (I'd argue even none of those) must be solved by this extension and all of them *can* be solved outside the extension (maybe except credentials sharing but that that can't be solved by anything).

Sure, having an encrypted local storage is a nice thing but it only protects the user from a smaller subset of threats, adds complexity to the code and increases the possible attack surface, especially if it is developed _in-house_. It can possibly result in less secure situation than without encryption at all just by giving users potentially false sense of security.

I think it is OK to have unencrypted off-line cache as long as the user is properly informed about the risks and and implications. It's their data after all.

**Comment by bertrand-benoit** [comment]
The fact is as soon as there is a network failure (either Internet or Cloud access), the whole passwords database is empty, and it is no more possible to retrieve login/pwd for anything.

Is there any way to keep 'last version' of retrieved passwords even in this situation?

**Comment by jficz** [comment]
This might not be as easy as it sounds since most of the functionality (namely search, match, etc.) is now provided by the API if I read the code correctly. This means for the local cache to be usable in the same way it would be needed to reimplement at least a subset of the API logic inside extension as well.

Since imho the local cache is _very_ important in case we lose connection to NC for any reason, it should be implemented somehow. However, since it will only be used in "emergencies" I don't think we need a full feature parity with the on-line app. I for one would be happy with a simple fulltext search without any advanced features like matching with websites, autofill, etc.

**Comment by godfuture** [comment]
> This might not be as easy as it sounds since most of the functionality (namely search, match, etc.) is now provided by the API if I read the code correctly. This means for the local cache to be usable in the same way it would be needed to reimplement at least a subset of the API logic inside extension as well.
> 
> Since imho the local cache is _very_ important in case we lose connection to NC for any reason, it should be implemented somehow. However, since it will only be used in "emergencies" I don't think we need a full feature parity with the on-line app. I for one would be happy with a simple fulltext search without any advanced features like matching with websites, autofill, etc.
> 
> another option would be to (one-way) sync the passwords with native browser's password storage, if that is even possible

You got a point. How does the Android app cache the passwords? So many times I was happy to be able to fetch my credit card PIN from Android device offline...

**Comment by mrtumnus** [comment]
> This might not be as easy as it sounds since most of the functionality (namely search, match, etc.) is now provided by the API if I read the code correctly. This means for the local cache to be usable in the same way it would be needed to reimplement at least a subset of the API logic inside extension as well.
This explains why the app is so slow - it's querying the server!

> Since imho the local cache is _very_ important in case we lose connection to NC for any reason, it should be implemented somehow. However, since it will only be used in "emergencies" I don't think we need a full feature parity with the on-line app. I for one would be happy with a simple fulltext search without any advanced features like matching with websites, autofill, etc.
Agreed!

> another option would be to (one-way) sync the passwords with native browser's password storage, if that is even possible
Good idea.

Another thought is to use biometrics/password to encrypt the local cache (a la banking apps, etc).

**Comment by godfuture** [comment]
What is the current state of this ticket? As a workaround, I am using keepass in parallel which is not really satisfying.
